examples/fluids/src/misc.c

```diff
@@ -292,3 +286,23 @@ PetscErrorCode CreateMassQFunction(Ceed ceed, CeedInt N, CeedInt q_data_size, Ce
   CeedQFunctionAddOutput(*qf, "v", N, CEED_EVAL_INTERP);
   PetscFunctionReturn(0);
 }
+
+/* @brief L^2 Projection of a source FEM function to a target FEM space
+ *
+ * @param[in]  source_vec    Vec of the source FEM function. NULL indicates that rhs_matop_ctx->x_ceed already has source FEM function interpolated to
+ * quadrature points
+ * @param[out] target_vec    Vec of the target (result) FEM function
+ * @param[in]  rhs_matop_ctx MatopApplyContext for performing the RHS evaluation
+ * @param[in]  Minv          Vec of the inverted lumped mass matrix, or `NULL`
+ * @param[in]  ksp           KSP for solving the consistent projection problem, or `NULL`
+ */
+PetscErrorCode ComputeL2Projection(Vec source_vec, Vec target_vec, MatopApplyContext rhs_matop_ctx, Vec Minv, KSP ksp) {
+  PetscFunctionBeginUser;
+  if (!Minv && !ksp) SETERRQ(PetscObjectComm((PetscObject)source_vec), PETSC_ERR_ARG_INCOMP, "Minv and ksp may not be simultaneously NULL");
+  PetscCall(ApplyLocal_Ceed(source_vec, target_vec, rhs_matop_ctx));
+
+  if (Minv) PetscCall(VecPointwiseMult(target_vec, target_vec, Minv));
+  if (ksp) PetscCall(KSPSolve(ksp, target_vec, target_vec));
```

Comment:
I think we have a problem if *both* are passed. What if we get rid of the `Minv` argument and always call `KSPSolve`, but it's allowed to be configured as `ksp_type=preonly, pc_type=jacobi`, in which case it's the lumped projection? It goes through some stack frames, but the only problem-sized computation should be the `VecPointwiseMult` inside `PCApply_Jacobi`.

Reply:
I don't think it's a problem? At least when I initially did the turbulent statistics, I used a lumped mass matrix before then handing it off to KSP. Effectively, it just gives KSP a better initial guess. That's why I left it as waterfall rather than an if/else statement.

The only hesitance I have against removing the lumped mass matrix in favor of the KSP solver is that A) we'd have to compute the lumped mass matrix every time (possibly more efficient, but we'd probably want to test for that) and B) we may want to use lumped mass matrices other than row-sum in the future. I'm thinking specifically about "special lumping" lumping of Hughes, which might be more favorable for tetrahedral elements.

Reply:
Wait, I see what you mean. We can't pass `target_vec` as the RHS vector after applying the lumped mass matrix. That's a typo on my part.

Thoughts on keeping `Minv` in there while preventing the problem if both are passed?

Reply:
Alternatively, if we're disallowing both `Minv` and `ksp` to be passed simultaneously, does it make sense to just remove the `ComputeL2Projection` function altogether? It only replaces two functions calls (`ApplyLocal` + either `VecPointwiseMult` or `KSPSolve`) with one.

Reply:
Can you remind me of what "special lumping" is? Can it be done algebraically?

KSP reuses the preconditioner if the matrix hasn't changed.

Reply:
Here's the section in Hughes FEM book that describes special lumping:
![image](https://user-images.githubusercontent.com/20801821/219976731-185d4900-8168-483a-81a9-a433afa3f0e7.png)
 
 Not sure what exactly you mean by algebraically, but I think it would be straight forward enough to implement if/when the time comes.

Reply:
Thanks. I think we can revisit if necessary. It sounds like the special lumping technique must be applied at the element level, but I can imagine doing it globally. That is, compute the total volume of the domain as $V = 1^T M 1$ and compute the global diagonal $D$, then set the new diagonal as $d_{i} \gets d_i V / (\sum_j d_j)$. This variant can be implemented purely within `PCJacobi` and I'd expect it to have similar properties.

Reply:
Ok. So move to just `ksp` as input?

Reply:
That'd be my recommendation

Reply:
Ok, done now. Feel free to merge; not sure why the CI is failing on the HIP side of things.